Source/Core/VideoCommon/TextureCacheBase.cpp

```diff
@@ -438,48 +438,64 @@ void TextureCacheBase::SerializeTexture(AbstractTexture* tex, const TextureConfi
   const bool skip_readback = p.GetMode() == PointerWrap::MODE_MEASURE;
   p.DoPOD(config);
 
-  std::vector<u8> texture_data;
   if (skip_readback || CheckReadbackTexture(config.width, config.height, config.format))
   {
-    // Save out each layer of the texture to the staging texture, and then
-    // append it onto the end of the vector. This gives us all the sub-images
-    // in one single buffer which can be written out to the save state.
+    // First, measure the amount of memory needed, then (if necessary) copy out.
+    u32 total_size = 0;
     for (u32 layer = 0; layer < config.layers; layer++)
     {
       for (u32 level = 0; level < config.levels; level++)
       {
         u32 level_width = std::max(config.width >> level, 1u);
         u32 level_height = std::max(config.height >> level, 1u);
-        auto rect = tex->GetConfig().GetMipRect(level);
-        if (!skip_readback)
+
+        u32 stride = AbstractTexture::CalculateStrideForFormat(config.format, level_width);
+        u32 size = stride * level_height;
+
+        total_size += size;
+      }
+    }
+
+    u8* texture_data = static_cast<u8*>(p.DoExternal(total_size));
+
+    if (!skip_readback)
+    {
+      // Save out each layer of the texture to the staging texture, and then
+      // append it onto the end of the region. This gives us all the sub-images
+      // in one single buffer which can be written out to the save state.
```

Comment:
This comment is a bit out of date now. It would be nice to replace the mention of "one single buffer" (this buffer doesn't actually exist anymore) with something saying that the reason we write directly into the `PointerWrap` is because we want to avoid allocating memory for performance reasons.